### PR TITLE
Couch URL cleanup

### PR DESCRIPTION
**Pull request by millerdev**

### CODE REVIEW

File: settingshelper.py [hunk 1]
```diff
@@ -70,39 +70,39 @@ def get_temp_file(self, suffix="", prefix="tmp"):
...
         }
     else:
-        return '%(http_method)s://%(server)s' % {
-            'http_method': http_method,
-            'server': server_root,
-        }
+        # admin party + empty username and auth_token
+        db_params = None
+    http_method = 'https' if use_https else 'http'
+    uri = '{http_method}://{server_root}/{dbname}'.format(**locals())
```
Comment (snopoke) [inline settingshelper.py]: no sure about this pattern of using `locals()`. Seem like it could lead to bugs down time line if someone renames somehting.

File: settingshelper.py [hunk 2]
```diff
@@ -70,39 +70,39 @@ def get_temp_file(self, suffix="", prefix="tmp"):
...
+class DbUri(six.text_type):
+    """DB URI string with params attribute
 
+    The `params` attribute is a dict to be passed as keyword arguments
+    to `couchdbkit.client.Database` as in `Database(**uri.params)`
+
+    This allows database parameters to be passed with the URL without
+    major refactoring to change all the places that reference the URL
+    expecting it to be a string. Historically the parameters were
+    encoded in the URL.
```
Comment (snopoke) [inline settingshelper.py]: seems like you've had to refactor most places anyway so not sure if this is much of a benefit.

File: settingshelper.py (hunk 1 above)
Comment (millerdev) [reply settingshelper.py]: [Python 3 has syntax for this](https://www.python.org/dev/peps/pep-0498/). I think it will become a common pattern once we switch.

File: settingshelper.py (hunk 2 above)
Comment (millerdev) [reply settingshelper.py]: It's true, I did change most places. However there are a few I found it hard to follow mentally. For example:

https://github.com/dimagi/commcare-hq/blob/ac91a36f2e3c547e053cb1ff171135a3c8fed97b/corehq/apps/cachehq/tests.py#L36-L38

Maybe that one could be removed? Or this one:

https://github.com/dimagi/commcare-hq/blob/dm/couch-url-cleanup/corehq/ex-submodules/dimagi/utils/management/commands/prime_views.py#L30-L33

Looks like it's doing some URL parsing there, but `t[0]` should not even be a URL? Is that a bug or did I missed something? And there's a bunch of different settings (`COUCH_DATABASE`, `COUCH_DATABASES`, `COUCHDB_DATABASES`, `EXTRA_COUCHDB_DATABASES`), some of which contain URLs. Confusing and hard to follow. Trying to change the type of what used to be a URL string made the scope of this change bigger than I wanted to tackle. It was easier to keep it a string and fix all the places that create `Database` objects.

Comment (millerdev) [reply settingshelper.py]: I took another stab at this and it turns out it's even more involved than I thought. Also, I think this whole PR is currently broken because `couchdbkit` depends on commcare-hq's `COUCHDB_DATABASES` setting (seriously??!), and I think will have to be updated because of that.

https://github.com/dimagi/couchdbkit/blob/cf98f473b753343d5c9404f07bb1f286b4c4affd/couchdbkit/ext/django/loading.py#L34

Comment (dannyroberts) [reply settingshelper.py]: > couchdbkit depends on commcare-hq's `COUCHDB_DATABASES` setting

I wouldn't have characterized it this way. The file you loaded is in `couchdbkit.ext.django`, where it `couchdbkit` puts its django-related extensions. The reason we have a settings variable called `COUCHDB_DATABASES` in commcare-hq is because setting django setting called `COUCHDB_DATABASES` is part of the couchdbkit-django interface. (It's relatively common for django libraries to have as part of their interface "passing something in" as a specific django setting, and `couchdbkit.ext.django` is basically a couchdbkit django library.)

Comment (dannyroberts) [reply settingshelper.py]: I just had a thought: what if along these lines (parsing the stuff out from the URL) we actually just change _couchdbkit_ to automatically parse server urls as they come in? That'll be the easiest place to make the change; not sure if it's unclean 🤔but it seems okay to me, especially the other alternatives we have at this point.

Comment (millerdev) [reply settingshelper.py]: I had previously assumed `couchdbkit.ext.django` was an extension we added to couchdbkit, but after a bit of blaming I see [it was added 10 years ago by the original author of couchdbkit](https://github.com/dimagi/couchdbkit/commit/fec2f591640b108ed4a13298b7cd44a9d3eaa0a5). Anyway, blaming isn't going to move us forward here. I had previously thought `COUCHDB_DATABASES` was a commcare-hq-specific setting, but now I see it is a couchdbkit setting that commcare-hq happens to also depend on in many places. The important thing is that the structure of `COUCHDB_DATABASES` expects passwords be encoded in the URL.

I think I agree with your idea: we should change couchdbkit to parse the url and extract the username and password prior to constructing a new `CouchDB` in the `Server`.

https://github.com/dimagi/couchdbkit/blob/998744b82681756ee8484d278e29d97575a3760b/couchdbkit/client.py#L108-L109

I'll revert my most recent changes and implement that instead.

Comment (dannyroberts) [reply settingshelper.py]: Awesome! Seems like this has been a deceptively annoying thing to solve with few false starts (I definitely tried for 40 minutes at some point a few weeks ago and gave up on the tack I was taking), but this new route sound promising, and I'm excited to hear you think so too. I hope that goes well!

**Comment by millerdev** [comment]
Closing in favor of https://github.com/dimagi/couchdbkit/pull/66
